promql/parser/node.go

```diff
@@ -0,0 +1,194 @@
+			return len(node.VectorMatching.MatchingLabels) > 0 || node.ReturnBool || node.VectorMatching.On
+		case *AggregateExpr:
+			return len(node.Grouping) > 0
+		}
+	case scalars:
+		if node, ok := n.currentNode.(*BinaryExpr); ok {
+			if node.LHS.Type() == ValueTypeScalar || node.RHS.Type() == ValueTypeScalar {
+				return true
+			}
+		}
+	case multiArguments:
+		if node, ok := n.currentNode.(*Call); ok {
+			return len(node.Args) > 1
+		}
+	case aggregateParent:
+		if len(n.ancestors) < 2 {
+			return false
+		}
+		if _, ok := n.ancestors[len(n.ancestors)-2].(*AggregateExpr); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// nodeHistory returns the ancestors of the node the item position range is passed of,
+// along with the node in which the item is present. This is done with the help of AST.
+// posRange can also be called as itemPosRange since carries the position range of the lexical item.
+func (n *nodeInfo) nodeHistory(head Node, posRange PositionRange, stack []Node) ([]Node, Node) {
+	if head.PositionRange().Contains(posRange) {
+		stack = append(stack, head)
+	}
+	for _, child := range Children(head) {
+		if child.PositionRange().Contains(posRange) {
+			return n.nodeHistory(child, posRange, stack)
+		}
+	}
+	return stack, head
+}
+
+// reduceNonNewLineExprs reduces those expressions from the history that are not
+// necessary for base indent. Base indent is applied on those items that have the
+// tendency to fall on a new line. However, items like Left_Bracket, Right_Bracket
+// never fall on new line. This means that when they (node in which they are present)
+// will be encountered, the stack already has MatrixSelector or SubQuery in it, but
+// their very existence harms the base indent. This is because they never contribute
+// to the indent, meaning the (metric_name[5m]) will get base indent as 3 (ParenExpr,
+// MatrixSelector, VectorSelector) but it should actually be 2 according to the requirements.
+// Hence, these unwanted expressions that do not contribute to the base indent
+// should be reduced.
+func reduceNonNewLineExprs(history []Node) []Node {
+	var temp []Node
+	if !(len(history) > 1) {
+		return history
+	}
+	for i := range history {
+		if !(isNodeType(history[i], matrixExpr) || isNodeType(history[i], subQueryExpr)) {
+			temp = append(temp, history[i])
+		}
+	}
+	return reduceBinary(temp)
+}
+
+// reduceBinary reduces from end, the continuous occurrence
+// of binary expression to its single representation.
+func reduceBinary(history []Node) []Node {
+	var temp []Node
+	if !(len(history) > 1) {
+		return history
+	}
+	for i := 0; i < len(history)-1; i++ {
+		if !(isNodeType(history[i], binaryExpr) && isNodeType(history[i+1], binaryExpr)) {
+			temp = append(temp, history[i])
+		}
+	}
+	temp = append(temp, history[len(history)-1])
+	return temp
+}
+
+func isNodeType(node Node, typ uint) bool {
+	switch typ {
+	case aggregateExpr:
+		if _, ok := node.(*AggregateExpr); ok {
+			return true
+		}
+		return false
```

Comment:
This line is redundant in all the `case`. The `return false` at the end is enough.